src/Circle.cpp

```diff
@@ -0,0 +1,70 @@
+//
+// Created by emiel on 24-02-20.
+//
+
+#include "Circle.h"
+
+namespace rtt {
+
+bool Circle::doesIntersectOrContain(const Vector2 &other) { return std::abs((center - other).length()) < radius; }
+
+bool Circle::doesIntersectOrContain(const Line &other) { return other.distanceToLine(center) < radius; }
+
+bool Circle::doesIntersectOrContain(const LineSegment &other) { return other.distanceToLine(center) < radius; }
+
+bool Circle::doesIntersectOrContain(const Circle &other) { return std::abs((center - other.center).length()) < (radius + other.radius); }
+
+bool Circle::doesIntersectOrContain(const Rectangle &other) {
+    // https://stackoverflow.com/questions/401847/circle-rectangle-collision-detection-intersection
+
+    Vector2 rectCenter = other.center();
+    Vector2 distanceToCircle;
+
+    double rectWidth = other.width();
+    double rectHeight = other.height();
+
+    distanceToCircle.x = abs(center.x - rectCenter.x);
+    distanceToCircle.y = abs(center.y - rectCenter.y);
+
+    if (distanceToCircle.x > (rectWidth / 2 + radius)) return false;
+    if (distanceToCircle.y > (rectHeight / 2 + radius)) return false;
+
+    if (distanceToCircle.x <= (rectWidth / 2)) return true;
+    if (distanceToCircle.y <= (rectHeight / 2)) return true;
+
+    double cornerDistance_sq = std::pow((distanceToCircle.x - rectWidth / 2), 2) + std::pow((distanceToCircle.y - rectHeight / 2), 2);
+
+    return (cornerDistance_sq <= (radius * radius));
+}
+
+bool Circle::doesIntersectOrContain2(const Rectangle &other) {
+    // https://stackoverflow.com/questions/401847/circle-rectangle-collision-detection-intersection
+
+    if (other.contains(center)) return true;
+
+    std::vector<LineSegment> segments = other.lines();
+    for (const LineSegment &segment : segments)
+        if (doesIntersectOrContain(segment)) return true;
+
+    return false;
+}
+
+Vector2 Circle::project(const Vector2 &point) { return center + (point - center).stretchToLength(radius); }
```

Comment:
This will go wrong if the circle has negative radius (e.g. the line will end up at the other side of the circle)

Reply:
There's also no unit test for this.

Reply:
Positive radii are enforced through the constructors. A circle can only have a negative radius if someone explicitly assigns it a negative radius, for which I assume they have a good reason. There are unit tests for the project function.